File: go/0039_sum_combination.go
package main

func combinationSum(candidates []int, target int) [][]int {
	var res [][]int
	for i, cand := range(candidates) {
		helper(target, i, candidates[i], []int{cand}, &res, candidates)
	}
	return res
}

func helper(target, curIndex, curSum int, cur []int, res *[][]int, candidates []int) {
	if curSum == target {
		ans := make([]int, len(cur))
		copy(ans, cur)
		*res = append(*res, ans)
	} else if curSum < target {
		for i := curIndex; i < len(candidates); i++ {
			cur = append(cur, candidates[i])
			helper(target, i, curSum + candidates[i], cur, res, candidates)
			cur = cur[:len(cur) - 1]
		}
	}
}

File: python/0039_sum_combination.py
class Solution:
    def combinationSum(self, candidates, target):
        dp = [[] for x in range(target+1)]
        for candidate in candidates:
            if candidate <= target:
                dp[candidate].append([candidate])
 
        for i, iSumsList in enumerate(dp):
            for candidate in candidates:
                if i + candidate > target:
                    continue
                    
                for iSums in iSumsList:
                    if iSums[-1] <= candidate:
                        dp[i+candidate].append(iSums + [candidate])
        return dp[target]
        